plugins/BEdita/API/src/Middleware/AnalyticsMiddleware.php

```diff
@@ -0,0 +1,170 @@
+            Log::config('analytics', [
+                'className' => 'File',
+                'path' => LOGS,
+                'scopes' => ['analytics'],
+                'file' => 'analytics',
+            ]);
+        }
+    }
+
+    /**
+     * Getter for $startTime
+     *
+     * @return float
+     * @codeCoverageIgnore
+     */
+    public function getStartTime()
+    {
+        return $this->startTime;
+    }
+
+    /**
+     * Getter for $data
+     *
+     * @return float
+     * @codeCoverageIgnore
+     */
+    public function getData()
+    {
+        return $this->data;
+    }
+
+    /**
+     * Register new analytics extension callback
+     *
+     * @param mixed $newCallback Callback to register
+     * @return void
+     */
+    public static function registerCallback($newCallback)
+    {
+        if (!is_callable($newCallback)) {
+            Log::warning('Bad callback ' . print_r($newCallback, true));
+
+            return;
+        }
+        static::$callbacks[] = $newCallback;
+    }
+
+    /**
+     * Get analytics custom data from registered callbacks
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @return array
+     */
+    protected function readCallbackData(ServerRequestInterface $request, ResponseInterface $response)
+    {
+        if (empty(static::$callbacks)) {
+            return [];
+        }
+
+        $res = [];
+        $params = [$request, $response];
+        foreach (static::$callbacks as $call) {
+            try {
+                $values = call_user_func_array($call, $params);
+                $res[] = $values;
+            } catch (Throwable $t) {
+                Log::error('Error calling callback ' . print_r($call, true));
+            }
+        }
+
+        return $res;
+    }
+
+    /**
+     * The middleware action.
+     *
+     * @param \Psr\Http\Message\ServerRequestInterface $request The request.
+     * @param \Psr\Http\Message\ResponseInterface $response The response.
+     * @param callable $next The next middleware to call.
+     * @return \Psr\Http\Message\ResponseInterface A response.
+     */
+    public function __invoke(ServerRequestInterface $request, ResponseInterface $response, callable $next)
+    {
+        $response = $next($request, $response);
+
+        /**
+         * TODO: add custom application error code
+         */
+        $this->data = [
+            'r' => $request->getEnv('REQUEST_TIME'),
+            'a' => CurrentApplication::getApplicationId(),
+            'usr' => LoggedUser::id(),
+            'm' => $request->getMethod(),
+            'url' => $request->getUri()->getPath(),
+            'q' => $request->getUri()->getQuery(),
+            's' => $response->getStatusCode(),
+            'x' => $this->readCallBackData($request, $response),
+            'e' => round(microtime(true) - $this->startTime, 4, PHP_ROUND_HALF_EVEN),
```

Comment:
If this delta-time computation is made here, it is not completely clear (at least to me) whether execution times from things like `$this->readCallBackData()` are taken into account — it is an implementation detail the order in which the interpreter processes the instructions in lines 155-163.

To clear out clouds I'd suggest to either compute the total execution time _before_ (at line 153), or _after_ (at line 165) then adding it to `$this->data`.

-------------------

By the way, the interpreter seems to process instructions in the most intuitive order:

```php
<?php

function a()
{
    echo 'A' . PHP_EOL;

    return 'A';
}

function b()
{
    echo 'B' . PHP_EOL;

    return 'B';
}

$data = [
    'a' => a(),
    'b' => b(),
];

var_dump($data);

/****** OUTPUT ******
A
B
array(2) {
  ["a"]=>
  string(1) "A"
  ["b"]=>
  string(1) "B"
}
********************/
```

However, I'd prefer to make it as clear as possible.

Reply:
Agreed, let's make it more clear